File: sections/28_dp_i_mixed_contest/problems/c_prerequisite_tree/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, budget;
    if (!(cin >> n >> budget)) return 0;
    vector<int> need(n);
    vector<long long> value(n);
    for (int i = 0; i < n; i++) cin >> need[i] >> value[i];
    vector<vector<int>> children(n);
    for (int child = 1; child < n; child++) {
        int parent;
        cin >> parent;
        children[parent - 1].push_back(child);
    }

    vector<int> order = {0};
    for (int i = 0; i < (int)order.size(); i++) {
        int u = order[i];
        for (int v : children[u]) order.push_back(v);
    }

    const long long NEG = -(1LL << 60);
    vector<vector<long long>> dp(n, vector<long long>(budget + 1, NEG));
    for (int idx = n - 1; idx >= 0; idx--) {
        int u = order[idx];
        if (need[u] <= budget) dp[u][need[u]] = value[u];
        for (int v : children[u]) {
            vector<long long> merged = dp[u];
            for (int used = 0; used <= budget; used++) {
                if (dp[u][used] == NEG) continue;
                for (int add = 1; used + add <= budget; add++) {
                    if (dp[v][add] == NEG) continue;
                    merged[used + add] = max(merged[used + add], dp[u][used] + dp[v][add]);
                }
            }
            dp[u].swap(merged);
        }
    }

    long long answer = 0;
    for (long long x : dp[0]) answer = max(answer, x);
    cout << answer << '\n';
    return 0;
}

File: sections/28_dp_i_mixed_contest/problems/c_prerequisite_tree/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    budget = next(it)
    time = [0] * n
    value = [0] * n
    for i in range(n):
        time[i] = next(it)
        value[i] = next(it)
    children = [[] for _ in range(n)]
    for child in range(1, n):
        parent = next(it) - 1
        children[parent].append(child)

    order = [0]
    for u in order:
        order.extend(children[u])

    neg = -10**30
    dp = [[neg] * (budget + 1) for _ in range(n)]
    for u in reversed(order):
        if time[u] <= budget:
            dp[u][time[u]] = value[u]
        for v in children[u]:
            merged = dp[u][:]
            for used in range(budget + 1):
                if dp[u][used] <= neg // 2:
                    continue
                for add in range(1, budget - used + 1):
                    if dp[v][add] <= neg // 2:
                        continue
                    candidate = dp[u][used] + dp[v][add]
                    if candidate > merged[used + add]:
                        merged[used + add] = candidate
            dp[u] = merged
    answer = max(0, max(dp[0]))
    print(answer)


if __name__ == "__main__":
    main()
